nntrainer/layers/time_dist.cpp

```diff
@@ -149,11 +193,85 @@ void TimeDistLayer::setDistLayer(std::shared_ptr<Layer> l) {
 };
 
 void TimeDistLayer::calcDerivative() {
-  // NYI
+  Tensor &derivative_ = net_hidden[0]->getGradientRef();
+  Tensor &hval_ = net_hidden[0]->getVariableRef();
+  Tensor &ret_ = net_input[0]->getGradientRef();
+  Tensor &input_ = net_input[0]->getVariableRef();
+
+  TensorDim der_dim = derivative_.getDim();
+  TensorDim ret_dim = ret_.getDim();
+
+  TensorDim r_dim = {ret_dim[2], 1, 1, ret_dim[3]};
+  TensorDim d_dim = {der_dim[2], 1, 1, der_dim[3]};
+
+  for (unsigned int i = 0; i < der_dim[0]; ++i) {
+    Tensor ret_iter =
+      ret_.getSharedDataTensor(r_dim, i * r_dim.batch() * r_dim.width());
+    Tensor in_iter =
+      input_.getSharedDataTensor(r_dim, i * r_dim.batch() * r_dim.width());
+    Tensor d_iter =
+      derivative_.getSharedDataTensor(d_dim, i * d_dim.batch() * d_dim.width());
+    Tensor hval_iter =
+      hval_.getSharedDataTensor(d_dim, i * d_dim.batch() * d_dim.width());
+
+    Var_Grad in_var(r_dim, true, false, dist_layer->getName() + ":input");
+    Var_Grad out_var(d_dim, true, false, dist_layer->getName() + ":output");
+
+    in_var.initializeGradient(ret_iter);
+    in_var.initializeVariable(in_iter);
+    out_var.initializeGradient(d_iter);
+    out_var.initializeVariable(hval_iter);
+
+    dist_layer->setInputBuffers({std::make_shared<Var_Grad>(in_var)});
+    dist_layer->setOutputBuffers({std::make_shared<Var_Grad>(out_var)});
+
+    dist_layer->calcDerivative();
+  }
+
+  ret_.copy(transposeTensor(ret_));
+  // We are not going to transpose the date. The Date is not used anymore.
+  // It will be overwritten at next iteration
+  // Just reshpae the tensors
+  hval_.reshape({der_dim[2], 1, der_dim[0], der_dim[3]});
+  derivative_.reshape({der_dim[2], 1, der_dim[0], der_dim[3]});
+  input_.reshape({ret_dim[2], 1, ret_dim[0], ret_dim[3]});
 }
 
 void TimeDistLayer::calcGradient() {
-  // NYI
+  // Even if the dist_layer->getNumWeights() == 0, We do transpose here
+  // for the calculation of derivatives and overwrite original tensors.
+  // And use them in calcDerivatives() without transpose.
+  transposeInOut();
+
+  if (dist_layer->getNumWeights() == 0)
+    return;
+
+  Tensor &input_ = net_input[0]->getVariableRef();
+  Tensor &derivative_ = net_hidden[0]->getGradientRef();
+
+  TensorDim der_dim = derivative_.getDim();
+  TensorDim in_dim = input_.getDim();
+
+  TensorDim i_dim = {in_dim[2], 1, 1, in_dim[3]};
+  TensorDim d_dim = {der_dim[2], 1, 1, der_dim[3]};
+
+  for (unsigned int i = 0; i < der_dim[0]; ++i) {
+    Tensor in_iter =
+      input_.getSharedDataTensor(i_dim, i * i_dim.batch() * i_dim.width());
+    Tensor d_iter =
+      derivative_.getSharedDataTensor(d_dim, i * d_dim.batch() * d_dim.width());
+
+    Var_Grad in_var(i_dim, true, false, dist_layer->getName() + ":input");
+    Var_Grad out_var(d_dim, true, false, dist_layer->getName() + ":output");
```

Comment:
How about making this make_shared here?
Current implementation will copy at line 270~271.

Reply:
As you can see, setInputBuffers(inputs) use pointer. Not copy the data.
```
net_input = inputs
```

Reply:
Yes, I though copying was happening here. Thank you for the clarification :)